### ISSUE TITLE
File Creation Events 

### ISSUE THREAD

**Issue by ghost** [eclipse-che/che#4220]
I would like to know more about how events are distributed around Che, and how it goes from workspace to ide and the other way around.

For example
I would like to have notification in IDE when some file is created.

So, when FILE.txt is created, at some path, event is sent to the IDE. And on ide side I have access to path of that file. 
Which event would that be? And how it should be implemented on both sides, server and client.

What I would like is to have something similar to ProjectListeners.java at plugin-java plugin 
https://github.com/eclipse/che/blob/master/plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/ProjectListeners.java

So when some file is CREATED, the event is sent.

I've tried some things, but it hasn't worked for me, so I would like if someone could explain to me a bit more thorough. I also seen similar things answered, but it wasn't exactly the solution, and I couldn't make it work.

I would also like to know if it's possible to run _commands_ from server side of the application?

**Comment by ghost** [comment]
@grunf93 

Recently, we have introduced one limitation: events are fired only for files that are located in a directory that is expanded in a project tree.

You may want to take a look at server side things:

```
org.eclipse.che.api.vfs.watcher.FileWatcherManager
org.eclipse.che.api.core.jsonrpc.RequestTransmitter
org.eclipse.che.api.vfs.impl.file.event.detectors.ProjectTreeTracker
```

And client side classes:

```
org.eclipse.che.ide.api.event.ng.ProjectTreeStateNotificationOperation
org.eclipse.che.ide.part.explorer.project.ProjectExplorerPresenter#initFileWatchers
```

This limitation has been dictated by such use cases as running npm install, and node_modules directory being quickly populated with a huge amount of files. When firing events for each files, the IDE just goes unresponsive.

**Comment by ghost** [comment]
@eivantsov  

Hm, ok, then I have to do workaround. Which events/classes are responsible for opening files in IDE?
So I would like to doubleclick on file in tree, and to have an action based on that. 
I tried using 

`org.eclipse.che.ide.api.event.FileEvent.FileEventHandler`
But not sure if I am implementing it the wrong way. 

I implemented something very similar to 
`org.eclipse.che.ide.extension.ant.client.AntExtension`
but I use FileEvent.FileOperation.OPEN
but for my own usage, so when I open file I want, command is ran.
Maybe I haven't fiddled around with it enough, but I think it doesn't work.

**Comment by dkulieshov** [comment]
Hi, 
`org.eclipse.che.ide.extension.ant.client.AntExtension` is a good example, however if I understood your needs correctly I would recommend you to take a look on how it is done here: `org.eclipse.che.ide.api.event.ng.FileOpenCloseEventListener`.

Briefly, 
- we register a handler for `org.eclipse.che.ide.api.event.FileEvent`
- on _file open_ event we call _`newFileTrackingStartEvent`_ 
  - this eventually leads to sending a message to a server side (via JsonRPC) here `org.eclipse.che.ide.api.event.ng.ClientServerEventService` 
  - this starts server-side file tracking
- on _file close_ event we call _`newFileTrackingStopEvent`_
  - similarly this leads to sending message to stop sever-side file tracking
`org.eclipse.che.ide.api.event.ng.EditorFileStatusNotificationOperation` this class is used to handle notification from sever-side.

I assume you're moving in the right direction but to my greatest regret I can't help you more until you show me the code snippet that you're experiencing problems with.

**Comment by ghost** [comment]
@grunf93 does @dkuleshov's response help?

**Comment by ghost** [comment]
Hi @eivantsov  and @dkuleshov 

I am sorry, I didn't have access to what I'm working on during the weekend. For now it's helpful, but I will check these things a bit later in the day.

Thanks!

**Comment by ghost** [comment]
@eivantsov  and @dkuleshov 
Unfortunately, I didn't have time to test it today, since I started doing workaround of what I am trying to accomplish. But yet I got stuck once again.

I hope I ain't bothering you too much but I am trying as much as possible to understand how Eclipse Che works. 
Is there any file implementation or class, where I could give Path to the file as an argument and it would return me a VirtualFile which can be opened through EditorAgentImpl class in editor. 

Is ```org.eclipse.che.ide.resources.impl.FileImpl.java``` what I'm looking for

I know it might look a bit chaotic what I'm trying to do, but as Eclipse Che is the first big project I am doing, it is a bit hard for me.

This was kind of my workflow.
1) First I tried to make it run as an event between Workspace and Client, it haven't worked for me, maybe because I used older version of Che (5.0.1), or maybe because I haven't used correct classes. Anyway, having an event with turned off IDE, wouldn't work for me, so I abandoned the idea.

2) Idea was to have action occur by clicking on the file in the project tree, and by opening it, it would trigger some actions, but as I said I didn't have time to test it today because I switched to third idea and started implementation of it.

3) To have specific buttons in my own specific part of IDE, which would trigger appropriate actions. I got Paths from the workspace, so I just need to open editor / trigger actions I need. And that's why I need to open file in editor. 

Thanks a lot for your help!

**Comment by dkulieshov** [comment]
@grunf93 As I understand you're working with old (5.0.1) version of Che, so before we proceed I would recommend you to test your solutions with the most actual 5.4.0 version as there were done lots of changes and it is quite possible that your ideas would be easier to implement with the fresh version.

**Comment by ghost** [comment]
@dkuleshov Sorry for not mentioning it, I upgraded to 5.3.0, as it was newest at the time. Is it really necessary to update it further?

**Comment by grunf93** [comment]
This is the snippet I tried using
			
    Path testPath = Path.valueOf("/projects/testFile123.txt");
    appContext.getWorkspaceRoot().getFile(testPath).then(new Operation<Optional<File>>() {
		@Override
		public void apply(Optional<File> arg) throws OperationException {
						editorAgent.openEditor(arg.get());
					}
		}); 

I tried using NotificationManager.notify and I get that the file is absent(), but I can see the file from the terminal.
Is this the correct way to open the file from code via path? For me it is not working. 

I am unable to debug it via superdev mode since Che moved to GWT 2.8.0 (it is not updated in documentation at the site) as I am getting this 
`    internal compiler exception occurred com.google.gwt.dev.jjs.InternalCompilerException: Error constructing Java AST`

As I read this has something to do with Eclipse JDT, so I guess the error might be on my side?
